XenAdmin/SettingsPanels/NRPEEditPage.cs

```diff
@@ -0,0 +1,397 @@
+            return true;
+        }
+
+        private string AllowHostsWithoutLocalAddress(string allowHosts)
+        {
+            string UpdatedAllowHosts = "";
+            string[] AllowHostArray = allowHosts.Split(',');
+            foreach (string allowHost in AllowHostArray)
+            {
+                if (!allowHost.Trim().Equals("127.0.0.1") &&
+                    !allowHost.Trim().Equals("::1"))
+                {
+                    UpdatedAllowHosts += allowHost + ",";
+                }
+            }
+            return UpdatedAllowHosts.Length == 0 ? NRPEHostConfiguration.ALLOW_HOSTS_PLACE_HOLDER :
+                UpdatedAllowHosts.Substring(0, UpdatedAllowHosts.Length - 1);
+        }
+
+        private void UpdateOtherComponentBasedEnableNRPECheckBox(bool check)
+        {
+            if (check)
+            {
+                AllowHostsTextBox.Enabled = true;
+                AllowHostsLabel.Enabled = true;
+                DebugLogCheckBox.Enabled = true;
+                SslDebugLogCheckBox.Enabled = true;
+                CheckDataGridView.Enabled = true;
+                CheckDataGridView.BackgroundColor = Color.FromKnownColor(KnownColor.Window);
+                CheckDataGridView.DefaultCellStyle.BackColor = Color.FromKnownColor(KnownColor.Window);
+            }
+            else
+            {
+                AllowHostsTextBox.Enabled = false;
+                AllowHostsLabel.Enabled = false;
+                DebugLogCheckBox.Enabled = false;
+                SslDebugLogCheckBox.Enabled = false;
+                CheckDataGridView.Enabled = false;
+                CheckDataGridView.BackgroundColor = Color.FromKnownColor(KnownColor.Control);
+                CheckDataGridView.DefaultCellStyle.BackColor = Color.FromKnownColor(KnownColor.Control);
+            }
+        }
+
+        private void EnableNRPECheckBox_CheckedChanged(object sender, EventArgs e)
+        {
+            UpdateOtherComponentBasedEnableNRPECheckBox(EnableNRPECheckBox.Checked);
+        }
+
+        private void AllowHostsTextBox_GotFocus(object sender, EventArgs e)
+        {
+            if (AllowHostsTextBox.Text.Equals(NRPEHostConfiguration.ALLOW_HOSTS_PLACE_HOLDER))
+            {
+                AllowHostsTextBox.Text = "";
+                AllowHostsTextBox.ForeColor = Color.FromKnownColor(KnownColor.ControlText);
+            }
+        }
+
+        private void AllowHostsTextBox_LostFocus(object sender, EventArgs e)
+        {
+            if (string.IsNullOrWhiteSpace(AllowHostsTextBox.Text))
+            {
+                AllowHostsTextBox.Text = NRPEHostConfiguration.ALLOW_HOSTS_PLACE_HOLDER;
+                AllowHostsTextBox.ForeColor = Color.FromKnownColor(KnownColor.ControlDark);
+            }
+        }
+
+        private void CheckDataGridView_BeginEdit(object sender, DataGridViewCellCancelEventArgs e)
+        {
+            DataGridViewCell currentCell = CheckDataGridView.CurrentRow.Cells[e.ColumnIndex];

```

Comment:
Here and in the following event handler I'm not convinced that checking the colour is the right way to decide to use thewatermark or not. It would be bette if we had a way to check whether the user has modified the value or not, in which case the cell should have the normal colour to make it obvious.